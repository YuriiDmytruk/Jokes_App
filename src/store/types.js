export const ADD_JOKES = 'ADD_JOKES';
export const GET_JOKES = 'GET_JOKES';
