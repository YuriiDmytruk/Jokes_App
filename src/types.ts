export type Joke = {
  joke: string;
  category: string;
  id: number;
}

export type JokesState = {
  jokes: Joke[],
};

export type JokesActionAdd = {
  type: string,
  jokes?: Joke[],
};

export type JokeDelete = {
  type: string,
  id: number,
}

export type JokesActionFetch = {
  type: string,
  props: {
      jokesLastID: number,
      jokesAmount: number,
    }
};

export type JokesControlProps = {
  jokesLength: number;
  all: string;
  categories: string[];
  jokesLastID: number;
  setPage: (page: number) => void;
  setFilter: (filter: string) => void;
};

export type JokesPaginationProps = {
  setPage: (page: number) => void;
  jokesLength: number;
  JOKES_ON_PAGE: number;
  page: number;
};
