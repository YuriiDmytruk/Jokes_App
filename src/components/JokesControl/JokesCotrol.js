import React, { useState } from "react";
import { Button, Container, Form, Col, Row } from "react-bootstrap";

const componentMergin = { marginTop: "20px" };

export default function JokesControl(props) {
  const [amount, setAmount] = useState();
  const [filter, setFilter] = useState();
  const [isValid, setValid] = useState(false);

  const validateJokesAmountInput = (e) => {
    if (e.target.value.match(/^[0-9]+$/) != null) {
      const number = parseInt(e.target.value);
      if (number >= 1 && number <= 10) {
        setAmount(number);
        setValid(true);
        return;
      }
    }
    setValid(false);
  };

  const changeAmount = () => {
    props.fetchJokes(amount);
  };

  const changeFilter = () => {
    props.setFilter(filter);
  };

  return (
    <Container style={componentMergin}>
      <Row>
        <Col>
          <Row>
            <Col>
              <Form.Control
                type="text"
                id="inputPassword5"
                aria-describedby="passwordHelpBlock"
                onChange={(e) => {
                  validateJokesAmountInput(e);
                }}
              />
              <Form.Text muted>
                Enter number of jokes to add. Value must be 1-10
              </Form.Text>
            </Col>
            <Col md="auto">
              <Button disabled={!isValid} onClick={changeAmount}>
                Add
              </Button>
            </Col>
          </Row>
        </Col>
        <Col>
          <Row>
            <Col>
              <Form.Select
                aria-label="Default select example"
                onChange={(e) => {
                  setFilter(e.target.value);
                }}
              >
                {[props.all].concat(props.categories).map((e, i) => (
                  <option key={i} value={e}>
                    {e}
                  </option>
                ))}
              </Form.Select>
              <Form.Text muted>Choose joke category</Form.Text>
            </Col>
            <Col md="auto">
              <Button onClick={changeFilter}>Filter</Button>
            </Col>
          </Row>
        </Col>
      </Row>
    </Container>
  );
}
