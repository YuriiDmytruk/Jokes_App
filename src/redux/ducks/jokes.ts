import { Joke, JokesState, JokesActionFetch, JokesActionAdd, JokeDelete } from '../../types'
import { put, takeEvery, call } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';

import { fetchJokes as fetchJokesFromApi } from '../../api';

export const ADD_JOKES = 'ADD_JOKES';
export const FETCH_JOKES = 'FETCH_JOKES';
export const SET_JOKES = 'SET_JOKES';
export const DELETE_JOKE = 'DELETE_JOKE'

export const JOKES = 'JOKES'

const defaultState: JokesState = {
  jokes: JSON.parse(localStorage.getItem(JOKES)!)
};

export const jokesReducer = (
  state: JokesState = defaultState,
  action: any
): JokesState => {
  switch (action.type) {
    case ADD_JOKES:
      return { ...state, jokes: state.jokes.concat(action.jokes || []) };
    case SET_JOKES:
      return { ...state, jokes: action.jokes };
    case DELETE_JOKE:
      return { ...state, jokes: state.jokes.filter(joke => joke.id !== action.id) };
    default:
      return state;
  }
};

export const deleteJoke = (id: number): JokeDelete => {
  return {type: DELETE_JOKE, id: id}
}

export const setJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: SET_JOKES, jokes: jokes };
}

export const addJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: ADD_JOKES, jokes: jokes };
}

export const fetchJokes = (
  jokesLastID: number,
  jokesAmount: number
): JokesActionFetch => {
  return {
    type: FETCH_JOKES,
    props: {
      jokesLastID: jokesLastID,
      jokesAmount: jokesAmount,
    }
  };
}

export function* fetchJokesWorker(action: JokesActionFetch): SagaIterator {
  const data: Joke[] = yield call(
    fetchJokesFromApi,
    action.props.jokesLastID,
    action.props.jokesAmount
  );
  yield put(addJokes(data));
}

export function* jokesWatcher() {
  yield takeEvery(FETCH_JOKES, fetchJokesWorker);
}