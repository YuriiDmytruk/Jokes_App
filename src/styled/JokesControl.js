import styled from 'styled-components';

export const ComponentMergin = styled.div`
  margin-top: 20px;
`;

export const LoaderConteiner = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`;
