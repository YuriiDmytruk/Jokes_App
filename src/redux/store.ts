import { createStore, applyMiddleware, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware, { Task } from 'redux-saga';
import { jokesWatcher, jokesReducer } from './ducks/jokes';

import { JokesState } from '../types';

const sagaMiddleware = createSagaMiddleware();

export const store: Store<JokesState> = createStore(
  jokesReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(jokesWatcher as () => IterableIterator<Task>);