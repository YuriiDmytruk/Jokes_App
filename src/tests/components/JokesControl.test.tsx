import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { jokesReducer } from '../../redux/ducks/jokes';
import JokesControl from '../../components/JokesCotrol';
import { JokesControlProps } from '../../types';

test('should render JokesControl component and match snapshot', () => {
  const store = createStore(jokesReducer, { jokes: [] });

  const props: JokesControlProps = {
    jokesLength: 0,
    all: 'ALL',
    categories: [],
    jokesLastID: 0,
    setPage: (page: number) => {},
    setFilter: (filter: string) => {},
  };

  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <JokesControl
          jokesLength={props.jokesLength}
          all={props.all}
          categories={props.categories}
          jokesLastID={props.jokesLastID}
          setPage={props.setPage}
          setFilter={props.setFilter}
        />
      </MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
