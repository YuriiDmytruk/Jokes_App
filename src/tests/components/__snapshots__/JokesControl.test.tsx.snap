// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders JokesControl correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beyTiQ jLQmOh"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <input
                class="form-control"
                id="inputAmount"
                type="text"
                value=""
              />
              <small
                class="form-text text-muted"
              >
                Enter the number of jokes to add. Value must be 1-10.
              </small>
            </div>
            <div
              class="col-md-auto"
            >
              <button
                class="btn btn-primary"
                disabled=""
                type="button"
              >
                Add
              </button>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="sc-guDMob cqzFYr"
              />
            </div>
            <div
              class="col-md-auto"
            >
              <div
                class="dropdown"
              >
                <button
                  aria-expanded="false"
                  class="dropdown-toggle btn btn-primary"
                  type="button"
                >
                  ALL
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
